internal/app/grpcserver/grcserver.go

```diff
@@ -0,0 +1,195 @@
+	val, err := s.Repo.CreateOrGetFromStorage(ctx, req.Url, userID)
+	if err != nil {
+		var ue *models.UniqueError
+		if errors.As(err, &ue) {
+			isUniqueError = true
+		} else {
+			s.MyLogger.Debug("Error CreateOrGetFromStorage", zap.String("error CreateOrGetFromStorage", err.Error()))
+
+			return nil, fmt.Errorf("%w", status.Error(codes.InvalidArgument, err.Error()))
+		}
+	}
+
+	result := utils.ReturnURL(val, s.URLRepo)
+
+	return &pb.SaveURLResponse{
+		Result:        result,
+		IsUniqueError: isUniqueError,
+	}, nil
+}
+
+func (s *Server) GetURL(ctx context.Context, req *pb.GetURLRequest) (*pb.GetURLResponse, error) {
+	s.MyLogger.Debug("start SaveURL")
+
+	v, err := protovalidate.New()
+	if err != nil {
+		return nil, fmt.Errorf("%w", status.Error(codes.Internal, err.Error()))
+	}
+
+	if err = v.Validate(req); err != nil {
+		return nil, fmt.Errorf("%w", status.Error(codes.InvalidArgument, err.Error()))
+	}
+
+	val, err := s.Repo.GetOriginalURLFromStorage(ctx, req.Url)
+	if err != nil {
+		s.MyLogger.Debug("Error CreateOrGetFromStorage", zap.String("error CreateOrGetFromStorage", err.Error()))
+
+		return nil, fmt.Errorf("%w", status.Error(codes.InvalidArgument, err.Error()))
+	}
+
+	result := utils.ReturnURL(val, s.URLRepo)
+
+	return &pb.GetURLResponse{
+		Url: result,
+	}, nil
+}
+
+func (s *Server) GetStats(ctx context.Context, _ *emptypb.Empty) (*pb.GetStatsResponse, error) {
+	s.MyLogger.Debug("start SaveURL")
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+	stats, err := s.Repo.GetStatsFromStorage(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w", status.Error(codes.Internal, err.Error()))
+	}
+
+	return &pb.GetStatsResponse{
+		Urls:  int32(stats.URLs),
+		Users: int32(stats.Users),
+	}, nil
+}
+
+func (s *Server) SaveBatchURL(ctx context.Context, req *pb.SaveBatchURLRequest) (*pb.SaveBatchURLResponse, error) {
+	token, err := utils.GetTokenFromMetadata(ctx)
+
+	if err != nil {
+		return nil, fmt.Errorf("%w", status.Error(codes.Internal, err.Error()))
+	}
+
+	v, err := protovalidate.New()
+	if err != nil {
+		return nil, fmt.Errorf("%w", status.Error(codes.Internal, err.Error()))
+	}
+
+	if err = v.Validate(req); err != nil {
+		return nil, fmt.Errorf("%w", status.Error(codes.InvalidArgument, err.Error()))
+	}
+	_, _ = auth.GetUserID(token)
+
+	body, err := json.Marshal(req.UrlBatch)
+	if err != nil {
+		return nil, fmt.Errorf("%w", status.Error(codes.InvalidArgument, "invalid request"))
+	}
+
+	result, err := usecase.ProcessBatchURLs(ctx, body, s.Repo, token, s.URLRepo, s.MyLogger)
+	if err != nil {
+		return nil, fmt.Errorf("%w", status.Error(codes.Internal, err.Error()))
+	}
+	//nolint:asciicheck
+	r := сonvertToSaveBatchURLResponse(result)
+
+	s.MyLogger.Debug("start SaveURL")
+
+	return r, nil
+}
+
+func сonvertToSaveBatchURLResponse(batchURL models.BatchURL) *pb.SaveBatchURLResponse {
+	responseObjects := make([]*pb.SaveBatchURLResponse_URLObject, 0)
+
+	for _, event := range batchURL {
```

Comment:
It might be in the utils package

Reply:
move to utils, remane to Convert2BatchURLResponse